File: coding_int/chapter_2/2_8_Loop_Detection.py
class node:
    def __init__(self, value=None):
        self.next = None
        self.datavalue = value
    
    def __str__(self):
        return "node: {0}".format(self.datavalue)

class llist:
    def __init__(self, head=None):
        self.head = head
        self.tail = head

    def __iter__(self):
        current = self.head
        while(current):
            yield current
            current = current.next
    
    def __str__(self):
        current = self.head
        result = ""
        while(current):
            result = result + str(current.datavalue) + " "
            current = current.next
        return result

    def addValue(self, dataitem):

        if (isinstance(dataitem, node)):
            new_node = dataitem
        else:
            new_node = node(dataitem)
        if (self.head is None):
            self.head = new_node
            self.tail = new_node
        else:
            self.tail.next = new_node
            self.tail = new_node

def find_loop(linked_list):

    test_dict = {}

    for item in linked_list:
        if item in test_dict.keys():
            print("circular ref found")
            return item
        else:
            test_dict[item] = True
    print("all clear")
    return False


if __name__ == "__main__":

    test_list = llist()
    test_list.addValue(1)
    test_list.addValue(5)
    test_list.addValue(4)

    result = find_loop(test_list)
    print(result)

    a = node(4)
    b = node(6)
    c = node(8)

    a.next = b
    b.next = c
    c.next = a

    test_list.addValue(a)

    #for item in test_list:
    #    print(item)

    result = find_loop(test_list)
    print(result)


    

File: coding_int/chapter_2/2_8_Loop_Detection.cpp
#include<iostream>
#include<set>

class node {
    public:

    node *next;
    int datavalue;
};

class llist {

    private:

    node* head;
    node* tail;

    public:

    llist() {
        head=NULL;
        tail=NULL;
    }

    void addValue(int value){

        node *temp = new node;
        temp->datavalue = value;
        temp->next = NULL;

        if (head==NULL) {
            head = temp;
            tail = temp;
        } else {
            tail->next = temp;
            tail = temp;
        }
    }

    node* getHead() {
        return head;
    }

    void addNode(node *nodetoadd) {

        if (head==NULL) {
            head = nodetoadd;
            tail = nodetoadd;
        } else {
            tail->next = nodetoadd;
            tail = nodetoadd;
        }

    }

    void printList() {

        node *temp = new node;
        temp = head;

        while(temp) {
            std::cout << temp->datavalue << "\t";
            temp=temp->next;
        }
        std::cout << std::endl << "---end---" << std::endl;
    }

    int list_length() {

        int length  = 0;

        node *temp = new node;
        temp = head;
        while(temp) {
            ++length;
            temp = temp->next;
        }

        return length;
    }

    node *check_loop() {

        node *temp = new node;
        temp=head;

        std::set<node*> setOfNodes;

        while (temp) {
            if(setOfNodes.insert(temp).second == false) {
                std::cout << "loop found" << std::endl;
                return temp;
            } else {
                setOfNodes.insert(temp);
            } 
            temp=temp->next;
        }

        std::cout << "all clear" << std::endl;
        return temp;
    }
};

int main() {

    llist *main1 = new llist;

    main1->addValue(1);
    main1->addValue(3);
    main1->addValue(4);

    node *dupes = new node;
    dupes = main1->check_loop();

    node *n1 = new node;
    n1->datavalue = 8;

    node *n2 = new node;
    n2->datavalue = 12;

    node *n3 = new node;
    n3->datavalue = 41;

    n1->next = n2;
    n2->next = n3;
    n3->next = n1;

    main1->addNode(n1);
    dupes = main1->check_loop();



}